services/server/src/handlers/availabilitySlot.ts

```diff
@@ -0,0 +1,235 @@
+    slots: slotIds,
+  });
+
+  const result: IAvailabilitySlot.FindAvailabilitySlotsApiResponse = {
+    list: [
+      {
+        slots: paginatedSlots.list,
+        subslots: [
+          ...paginatedLessons.list.map((lesson) => ({
+            parent: lesson.slotId,
+            start: lesson.start,
+            end: dayjs(lesson.start)
+              .add(lesson.duration, "minutes")
+              .toISOString(),
+          })),
+          ...paginatedInterviews.list.map((interview) => ({
+            parent: interview.ids.slot,
+            start: interview.start,
+            end: dayjs(interview.start)
+              .add(INTERVIEW_DURATION, "minutes")
+              .toISOString(),
+          })),
+        ],
+      },
+    ],
+    total: paginatedSlots.total,
+  };
+
+  res.status(200).json(result);
+}
+
+/**
+ * This single endpoint is dedicated to create, update, and delete slots
+ * This is because when the user is preparing his schedule, he can create,
+ * delete, and update simultaneously.
+ */
+async function set(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  const allowed = isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+
+  const payload = setPayload.parse(req.body);
+
+  const creates = payload.slots.filter(
+    (obj) => obj.action === "create"
+  ) as Array<CreateAction>;
+  const updates = payload.slots.filter(
+    (obj) => obj.action === "update"
+  ) as Array<UpdateAction>;
+  const deletes = payload.slots.filter(
+    (obj) => obj.action === "delete"
+  ) as Array<DeleteAction>;
+
+  const error = await knex.transaction(async (tx) => {
+    const deleteIds = deletes.map((obj) => obj.id);
+    const ids = [...deleteIds, ...updates.map((obj) => obj.id)];
+
+    if (!isEmpty(ids)) {
+      const allExist = await availabilitySlots.allExist(ids, tx);
+      if (!allExist) return notfound.slot();
+
+      const isOwner = await availabilitySlots.isOwner({
+        slots: ids,
+        owner: user.id,
+        tx,
+      });
+      if (!isOwner) return forbidden();
+
+      // delete slots
+      if (!isEmpty(deletes)) {
+        const associatedLessons = await lessons.find({ slots: deleteIds, tx });
+        const associatedInterviews = await interviews.find({
+          slots: deleteIds,
+          tx,
+        });
+        const associatesCount =
+          associatedLessons.total + associatedInterviews.total;
+        if (associatesCount > 0) {
+          await lessons.cancelBatch({
+            ids: associatedLessons.list.map((l) => l.id),
+            canceledBy: user.id,
+            tx,
+          });
+          await interviews.cancelBatch({
+            ids: associatedInterviews.list.map((i) => i.ids.self),
+            canceledBy: user.id,
+            tx,
+          });
+          await availabilitySlots.markAsDeleted({ ids: deleteIds, tx });
+        } else {
+          await availabilitySlots.delete(deleteIds, tx);
+        }
+      }
+
```

Comment:
This logic has a lot of nested if/else which goes against clean code perenisables. I believe this will make it hard to mainating an extend it in the futue.